File: cpp/easy/GetIntersectionNode.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int a) : val(a), next(nullptr){};
};

ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
  int a_len = 0;
  int b_len = 0;
  ListNode *a_temp = headA;
  ListNode *b_temp = headB;
  while (a_temp || b_temp) {
    if (a_temp) {
      a_temp = a_temp->next;
      a_len++;
    }
    if (b_temp) {
      b_temp = b_temp->next;
      b_len++;
    }
  }
  if (a_len > b_len) {
    for (int i = 0; i < a_len - b_len; ++i) {
      headA = headA->next;
    }
  } else {
    for (int i = 0; i < b_len - a_len; ++i) {
      headB = headB->next;
    }
  }
  while (headA) {
    if (headA == headB)
      return headA;
    headA = headA->next;
    headB = headB->next;
  }
  return nullptr;
}

File: go/easy/GetIntersectionNode.go
package main

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	l1 := 0
	l2 := 0
	dummy1 := headA
	dummy2 := headB
	for dummy1 != nil {
		l1 = l1 + 1
		dummy1 = dummy1.Next
	}
	for dummy2 != nil {
		l2 = l2 + 1
		dummy2 = dummy2.Next
	}
	if l1 > l2 {
		for i := 0; i < l1-l2; i = i + 1 {
			headA = headA.Next
		}
	}
	if l1 < l2 {
		for i := 0; i < l2-l1; i = i + 1 {
			headB = headB.Next
		}
	}
	for headA != nil && headB != nil {
		if headA == headB {
			return headA
		}
		headA = headA.Next
		headB = headB.Next
	}
	return nil
}
